pysmt/smtlib/parser.py

```diff
@@ -855,44 +834,90 @@ def parse_atoms(self, tokens, command, min_size, max_size=None):
         if additional_token is not None:
             var = additional_token
         else:
             var = tokens.consume()
-        if var == "(":
+
+        res = None
+        if type_params and var in type_params:
+            return (var,) # This is a type parameter, it is handled recursively
+        elif var == "(":
             op = tokens.consume()
 
             if op == "Array":
                 idxtype = self.parse_type(tokens, command)
                 elemtype = self.parse_type(tokens, command)
                 self.consume_closing(tokens, command)
-                return ("Array", idxtype, elemtype)
+                res = self.env.type_manager.ArrayType(idxtype, elemtype)
 
-            if op != "_":
-                raise PysmtSyntaxError("Unexpected token '%s' in %s command." % \
-                                  (op, command))
-            ts = tokens.consume()
-            if ts != "BitVec":
-                raise PysmtSyntaxError("Unexpected token '%s' in %s command." % \
-                                  (ts, command))
+            elif op == "_":
+                ts = tokens.consume()
+                if ts != "BitVec":
+                    raise PysmtSyntaxError("Unexpected token '%s' in %s command." % \
+                                      (ts, command))
 
-            size = 0
-            dim = tokens.consume()
-            try:
-                size = int(dim)
-            except ValueError:
-                raise PysmtSyntaxError("Unexpected token '%s' in %s command." % \
-                                  (dim, command))
+                size = 0
+                dim = tokens.consume()
+                try:
+                    size = int(dim)
+                except ValueError:
+                    raise PysmtSyntaxError("Unexpected token '%s' in %s command." % \
+                                      (dim, command))
 
-            self.consume_closing(tokens, command)
-            return "BV%d" % size
+                self.consume_closing(tokens, command)
+                res = self.env.type_manager.BVType(size)
+            else:
+                #it must be a custom-defined type
+                base_type = self.cache.get(op)
+                if base_type is not None:
+                    if isinstance(base_type, _TypeDecl):
+                        pparams = []
+                        has_free_params = False
+                        for _ in range(base_type.arity):
+                            ty = self.parse_type(tokens, command, type_params=type_params)
+                            pparams.append(ty)
+                            if isinstance(ty, tuple):
+                                has_free_params = True
+                        if has_free_params:
+                            def definition(*args):
+                                params = []
+                                for x in pparams:
+                                    if isinstance(x, tuple):
```

Comment:
This not super clean, but it works. A principle dolution would be to represent the type-tree originating from a `define-sort` and provide a replace method once the types are instantiated